include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -18,43 +17,41 @@ namespace fft
 // ---------------------------------- Utils -----------------------------------------------
 
 template<typename SharedMemoryAccessor, typename Scalar>
-void exchangeValues(NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi, uint32_t threadID, uint32_t stride, NBL_REF_ARG(MemoryAdaptor<SharedMemoryAccessor>) sharedmemAdaptor)
+void exchangeValues(NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi, uint32_t threadID, uint32_t stride, NBL_REF_ARG(SharedMemoryAccessor) sharedmemAccessor)
 {
     const bool topHalf = bool(threadID & stride);
+    // Ternary won't take structs so we use this aux variable
     vector <Scalar, 2> toExchange = topHalf ? vector <Scalar, 2>(lo.real(), lo.imag()) : vector <Scalar, 2>(hi.real(), hi.imag());
-    shuffleXor<SharedMemoryAccessor, vector <Scalar, 2> >::__call(toExchange, stride, sharedmemAdaptor);
+    complex_t<Scalar> toExchangeComplex = {toExchange.x, toExchange.y};
+    shuffleXor<SharedMemoryAccessor, complex_t<Scalar> >::__call(toExchangeComplex, stride, sharedmemAccessor);
     if (topHalf)
-    {
-        lo.real(toExchange.x);
-        lo.imag(toExchange.y);
-    }
+        lo = toExchangeComplex;
     else
-    {
-        hi.real(toExchange.x);
-        hi.imag(toExchange.y);
-    }   
+        hi = toExchangeComplex;
 }
 
+} //namespace fft
+
 // ----------------------------------- End Utils -----------------------------------------------
 
 template<uint16_t ElementsPerInvocation, bool Inverse, typename Scalar, class device_capabilities=void>
 struct FFT;
 
 // For the FFT methods below, we assume:
 //      - Accessor is a global memory accessor to an array fitting 2 * _NBL_HLSL_WORKGROUP_SIZE_ elements of type complex_t<Scalar>, used to get inputs / set outputs of the FFT,
-//        that is, one "lo" and one "hi" complex numbers per thread, essentially 4 Scalars per thread. The data layout is assumed to be a whole array of real parts 
-//        followed by a whole array of imaginary parts. So it would be something like
-//        [x_0, x_1, ..., x_{2 * _NBL_HLSL_WORKGROUP_SIZE_}, y_0, y_1, ..., y_{2 * _NBL_HLSL_WORKGROUP_SIZE_}] 
-//      - SharedMemoryAccessor accesses a shared memory array that can fit _NBL_HLSL_WORKGROUP_SIZE_ elements of type complex_t<Scalar>, so 2 * _NBL_HLSL_WORKGROUP_SIZE_ Scalars 
+//        that is, one "lo" and one "hi" complex numbers per thread, essentially 4 Scalars per thread. 
+//        There are no assumptions on the data layout: we just require the accessor to provide get and set methods for complex_t<Scalar>.
+//      - SharedMemoryAccessor accesses a shared memory array that can fit _NBL_HLSL_WORKGROUP_SIZE_ elements of type complex_t<Scalar>, with get and set 
+//        methods for complex_t<Scalar>. It benefits from coalesced accesses   
```

Comment:
Shared Memory Accessor is still supposed to only have `uint32_t` methods because SMEM is usually banked in 32 banks